### DISCUSSION TITLE
Arbitrary channels (for multi-spectral and other uses)

### DISCUSSION THREAD

**Discussion by cesss** [ImageMagick/ImageMagick#3624] [Help]
Hi!

I've been searching for this and found no examples nor specific documentation, but it seems that ImageMagick supports a maximum of 32 channels. Is this right? The docs even tell you to set an "Undefined colorspace" for working with multi-spectral data, but no more info is available on that topic.

I ask because I'm working with physically-based rendering, performed with spectral samples instead of in RGB space, and I thought that ImageMagick could help at processing the spectral data before it's converted to RGB or sRGB for final display.

Do you know of any users working with multi-spectral images, or any examples online?

I suppose I could even use ImageMagick for converting N spectral channels into RGB, by defining the RGB weights of each spectral sample, couldn't I?

Then, what image format would you recommend? MIFF? Or maybe EXR? BTW, do you support 32bit EXR with an arbitrary number of channels, or is MIFF the only choice for what I'm describing? Is there any other format, apart from MIFF, capable of storing an arbitrary number of channels?

Then, I'm thinking in storing more data as arbitrary channels, such as UV coordinates, or even object IDs or material IDs (although I believe ImageMagick doesn't support heterogeneous types across channels --I mean, I believe they must be either all integer or all floating point, although AFAIK floating point can always preserve exact integers, so storing IDs in a floating point channel shouldn't hurt...)

Thanks!!

**Comment by snibgo** [comment]
> Do you know of any users working with multi-spectral images, or any examples online?

I don't.

TIFF can be used for arbitrary channels.

IM doesn't have command-line facilities for arbitrary channels. If you want N channels, you can use N/3 images each with 3 channels, but managing these is a pain.

For that, G'MIC is better. G'MIC can have any number of channels you want, provided the entire image fits into memory. For example: create an image width=12 pixels, height=34 pixels, with 1000 channels, where all pixels in the first channel are 255, and save the image as x.tiff:
```
gmic 12,34,1,1000,[255] output x.tiff
```
exiftool confirms the image dimensions, including the 1000 channels. ImageMagick can't successfully read x.tiff.

You might ask the question on https://discuss.pixls.us/ , where the main G'MIC author David Tschumperle hangs out, as do other G'MIC users, and some people there are at least interested in multi-spectral images.

**Reply by cesss** [reply]
Thanks a lot for your help, @snibgo 

Maybe, on second thought, I should consider writing my own imagebuffer small lib with only the features I need. I'm saying this because, while I don't want to reinvent the wheel, at the same time it seems there's no lib that fits what I need. Basically, what I want is a lib with a generic API for all image formats, that also provides a flexible configurable image buffer for drawing/editing them (by configurable I mean arbitrary channels, with arbitrary meanings, and even with different per-channel data type and depth), as well as color management support for performing the linear<>nonlinear conversion in a proper way. 

G'MIC is C++, and I prefer to choose C for external dependencies (mainly because C++ is changing every year, and C++ libraries usually adopt the changes and force you to update the compiler, which in turn makes you sometimes drop old systems just because you cannot install a new C++ compiler on them...  so I prefer C for my dependencies, that way I'm sure I won't face this kind of compatibility dilemmas in the future).

**Reply by fmw42** [reply]
IM does not yet support using more than 5 channels (CMYKA) in TIFF, because there has not yet been development for the read/write. Otherwise it could. You can do that with MIFF or MPC only right now.

**Comment by snibgo** [comment]
> Then, I'm thinking in storing more data as arbitrary channels, such as UV coordinates, or even object IDs or material IDs ...

A build of IM stores pixel values either as integers or as floating-point. I think TIFF files can be mixed, eg with some channels integer and others as floating-point. IM can't write mixed files. I don't know if IM or G'MIC can read such files.

> ... although AFAIK floating point can always preserve exact integers, ...

That is true only up to a certain limit. For example, 32-bit floats can store numbers up to 3.4e38, but most integers with more than 7 digits can't be recorded exactly.

> Maybe, on second thought, I should consider writing my own imagebuffer small lib with only the features I need.

I suspect that is a sensible approach.
